src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/SseParser_1.cs

```diff
@@ -0,0 +1,532 @@
+        private void ThrowIfNotFirstEnumeration()
+        {
+            if (Interlocked.Exchange(ref _used, 1) != 0)
+            {
+                throw new InvalidOperationException(SR.InvalidOperation_EnumerateOnlyOnce);
+            }
+        }
+
+        /// <summary>Reads data from the stream into the line buffer.</summary>
+        private int FillLineBuffer()
+        {
+            ShiftOrGrowLineBufferIfNecessary();
+
+            int offset = _lineOffset + _lineLength;
+            int bytesRead = _stream.Read(
+#if NET
+                _lineBuffer.AsSpan(offset));
+#else
+                _lineBuffer, offset, _lineBuffer.Length - offset);
+#endif
+
+            if (bytesRead <= 0)
+            {
+                _eof = true;
+                if (bytesRead < 0)
+                {
+                    throw new InvalidOperationException(SR.InvalidOperation_InvalidStreamNegativeBytes);
+                }
+            }
+
+            _lineLength += bytesRead;
+            return bytesRead;
+        }
+
+        /// <summary>Reads data asynchronously from the stream into the line buffer.</summary>
+        private async ValueTask<int> FillLineBufferAsync(CancellationToken cancellationToken)
+        {
+            ShiftOrGrowLineBufferIfNecessary();
+
+            int offset = _lineOffset + _lineLength;
+            int bytesRead = await
+#if NET
+                _stream.ReadAsync(_lineBuffer.AsMemory(offset), cancellationToken)
+#else
+                new ValueTask<int>(_stream.ReadAsync(_lineBuffer, offset, _lineBuffer.Length - offset, cancellationToken))
+#endif
+                .ConfigureAwait(false);
+
+            if (bytesRead <= 0)
+            {
+                _eof = true;
+                if (bytesRead < 0)
+                {
+                    throw new InvalidOperationException(SR.InvalidOperation_InvalidStreamNegativeBytes);
+                }
+            }
+
+            _lineLength += bytesRead;
+            return bytesRead;
+        }
+
+        /// <summary>Gets the UTF8 BOM.</summary>
+        private static ReadOnlySpan<byte> Utf8Bom => [0xEF, 0xBB, 0xBF];
+
+        /// <summary>Called at the beginning of processing to skip over an optional UTF8 byte order mark.</summary>
+        private void SkipBomIfPresent()
+        {
+            Debug.Assert(_lineOffset == 0, $"Expected _lineOffset == 0, got {_lineOffset}");
+
+            if (_lineBuffer.AsSpan(0, _lineLength).StartsWith(Utf8Bom))
+            {
+                _lineOffset += 3;
+                _lineLength -= 3;
+            }
+        }
+
+        /// <summary>Grows the buffer, returning the existing one to the ArrayPool and renting an ArrayPool replacement.</summary>
+        private static void GrowBuffer([NotNull] ref byte[]? buffer, int minimumLength)
+        {
+            byte[]? toReturn = buffer;
+            buffer = ArrayPool<byte>.Shared.Rent(Math.Max(minimumLength, DefaultArrayPoolRentSize));
```

Comment:
It's a bit concerning to me that we don't have a way to limit the line length, but I guess that's a concern for threat models and docs.